components/post_view/reaction/reaction.jsx

```diff
@@ -10,88 +10,152 @@ import OverlayTrigger from 'components/overlay_trigger';
-         */
         currentUserId: PropTypes.string.isRequired,
-
-        /*
-         * The name of the emoji for the reaction
-         */
         emojiName: PropTypes.string.isRequired,
-
-        /*
-         * The number of reactions to this post for this emoji
-         */
         reactionCount: PropTypes.number.isRequired,
-
-        /*
-         * Array of users who reacted to this post
-         */
         profiles: PropTypes.array.isRequired,
-
-        /*
-         * The number of users not in the profile list who have reacted with this emoji
-         */
         otherUsersCount: PropTypes.number.isRequired,
-
-        /*
-         * Array of reactions by user
-         */
         reactions: PropTypes.arrayOf(PropTypes.object).isRequired,
-
-        /*
-         * True if the user has the permission to add a reaction in this channel
-         */
         canAddReaction: PropTypes.bool.isRequired,
-
-        /*
-         * True if user has the permission to remove his own reactions in this channel
-         */
         canRemoveReaction: PropTypes.bool.isRequired,
-
-        /*
-         * The URL of the emoji image
-         */
         emojiImageUrl: PropTypes.string.isRequired,
-
         actions: PropTypes.shape({
-
-            /*
-             * Function to add a reaction to a post
-             */
             addReaction: PropTypes.func.isRequired,
-
-            /*
-             * Function to get non-loaded profiles by id
-             */
             getMissingProfilesByIds: PropTypes.func.isRequired,
-
-            /*
-             * Function to remove a reaction from a post
-             */
             removeReaction: PropTypes.func.isRequired,
         }),
         sortedUsers: PropTypes.object.isRequired,
     }
 
-    handleAddReaction = (e) => {
-        e.preventDefault();
-        const {actions, post, emojiName} = this.props;
-        actions.addReaction(post.id, emojiName);
+    constructor(props) {
+        super(props);
+
+        const {reactionCount, canRemoveReaction, canAddReaction} = this.props;
+        const {currentUserReacted} = this.props.sortedUsers;
+
+        if (currentUserReacted) {
+            this.state = {
+                userReacted: currentUserReacted,
+                reactedClass: 'Reaction reacted',
+                displayNumber: reactionCount,
+                reactedNumber: reactionCount,
+                unreactedNumber: reactionCount - 1,
+            };
+        } else {
+            this.state = {
+                userReacted: currentUserReacted,
+                reactedClass: 'Reaction unreacted',
+                displayNumber: reactionCount,
+                reactedNumber: reactionCount + 1,
+                unreactedNumber: reactionCount,
+            };
+        }
+
+        if (currentUserReacted && !canRemoveReaction) {
+            this.state = {
+                ...this.state,
+                reactedClass: 'Reaction read-only',
```

Comment:
Instead of storing the whole CSS class in the state, maybe it would make sense to derive it in the `render` method? "Reaction" will always be there and "read-only" is inferred by the props, so the only thing that really needs to be in state is the transitions like "reacting"

Reply:
Also can we use BEM for readability?

Reply:
I could, let me look at that again. The render method was really messy previously with all of the conditions. But I agree the parent class name doesn’t need to change.

Reply:
I was going to switch this back to BEM format and forgot. Will do that now.

Reply:
This should all match BEM conventions now, plus I'm only rendering the modifier classes now as the state updates.